### PR TITLE
Allow Nullable Foreign-Keys

### LINKED ISSUE #554: Support Nullable Foreign-Keys
A database I'm trying to connect to has a nullable foreign-key which odo doesn't currently handle:

**Test Case:**
```python
import blaze as bz
from numpy.random import randint
import sqlalchemy as sa


engine = sa.create_engine('sqlite://')
metadata = sa.MetaData(bind=engine)

T1 = sa.Table(
    'NullableForeignKeyDemo',
    metadata,
    sa.Column('pkid', sa.Integer, primary_key=True),
    sa.Column('label_id', sa.Integer, sa.ForeignKey("ForeignKeyLabels.pkid"), nullable=True),
)

T2 = sa.Table(
    'ForeignKeyLabels',
    metadata,
    sa.Column('pkid', sa.Integer, primary_key=True),
    sa.Column('label', sa.String),
)

metadata.create_all()

records1 = [
    {'pkid': idx, 'label_id': int(value)}
    for idx, value
    in enumerate(randint(0, 10, 10))
]
records1[-1]['label_id'] = None  # foreign-key is nullable!

records2 = [
    {'pkid': pkid, 'label': chr(pkid+65)}
    for pkid in range(10)
]
with engine.connect() as conn:
    conn.execute(T1.insert(), records1)
    conn.execute(T2.insert(), records2)

data = bz.data(T1)
```
As can be seen the inferred datashape doesn't recognize the map key as being nullable:
```python
In [42]: data
Out[42]: <'Table' data; _name='_4', dshape='var * {  pkid: int32,  label_id: map[int32, {pkid:...'>
```
...which then causes an issue when odo tries to convert `None` to an integer:
```python
In [43]: odo(data, pd.DataFrame)
Traceback (most recent call last):

  File "<ipython-input-43-9399f6fc5f8e>", line 1, in <module>
    odo(data, pd.DataFrame)

  File "C:\Miniconda3\lib\site-packages\odo\odo.py", line 91, in odo
    return into(target, source, **kwargs)

  File "C:\Miniconda3\lib\site-packages\multipledispatch\dispatcher.py", line 164, in __call__
    return func(*args, **kwargs)

  File "C:\Miniconda3\lib\site-packages\blaze\compute\core.py", line 372, in into
    return into(a, b.data, **kwargs)

  File "C:\Miniconda3\lib\site-packages\multipledispatch\dispatcher.py", line 164, in __call__
    return func(*args, **kwargs)

  File "C:\Miniconda3\lib\site-packages\odo\into.py", line 43, in wrapped
    return f(*args, **kwargs)

  File "C:\Miniconda3\lib\site-packages\odo\into.py", line 53, in into_type
    return convert(a, b, dshape=dshape, **kwargs)

  File "C:\Miniconda3\lib\site-packages\odo\core.py", line 83, in __call__
    return _transform(self.graph, *args, **kwargs)

  File "C:\Miniconda3\lib\site-packages\odo\core.py", line 106, in _transform
    x = f(x, excluded_edges=excluded_edges, **kwargs)

  File "C:\Miniconda3\lib\site-packages\odo\backends\sql.py", line 744, in select_or_selectable_to_frame
    dtype=[(str(c), dtypes[c]) for c in columns]))

TypeError: int() argument must be a string, a bytes-like object or a number, not 'NoneType'
```

### PR DESCRIPTION
**Pull request by dhirschfeld**
Fixes #554

### CODE REVIEW

File: odo/backends/sql.py [hunk 1]
```diff
@@ -225,7 +225,13 @@ def discover_sqlalchemy_selectable(t):
...
              for fkey in t.foreign_keys]
     for name, column in merge(*fkeys).items():
-        record[ordering[name]] = (name, column)
+        index = ordering[name]
+        _, key_type = record[index]
+        # If the foreign-key is nullable the column (map) key
+        # should be an Option type
+        if isinstance(key_type, Option):
+            column.key = Option(column.key)
+        record[index] = (name, column)
```
Comment (dhirschfeld) [inline odo/backends/sql.py]: This change ensures the map key is nullable if the foreign-key is nullable
```python
In [2]: data
Out[2]: <'Table' data; _name='_1', dshape='var * {  pkid: int32,  label_id: map[?int32, {pkid...'>
```

File: odo/backends/sql.py [hunk 2]
```diff
@@ -736,9 +742,15 @@ def select_or_selectable_to_frame(el, bind=None, dshape=None, **kwargs):
...
+                try:
+                    dtypes[field] = ty.to_numpy_dtype()
+                except TypeError:
+                    dtypes[field] = np.dtype(object)
         else:
-            dtypes[field] = dtype.to_numpy_dtype()
+            try:
+                dtypes[field] = dtype.to_numpy_dtype()
+            except TypeError:
+                dtypes[field] = np.dtype(object)
```
Comment (dhirschfeld) [inline odo/backends/sql.py]: This is also required to support nullable foreign-keys as there's no numpy equivalent of a nullable it.

This whole block seems a bit redundant to me though - if the goal is to push it into pandas why bother going through numpy at all? pandas is much looser with types than numpy so it seems a bit pointless raising an exception on conversion to numpy when pandas could handle the data perfectly well (just using an object dtype)

**Comment by dhirschfeld** [comment]
Without the second commit you get the below error:
```python
In [4]: data
Out[4]: <'Table' data; _name='_1', dshape='var * {  pkid: int32,  label_id: map[?int32, {pkid...'>

In [5]: odo(data, pd.DataFrame)
Traceback (most recent call last):

  File "<ipython-input-5-9399f6fc5f8e>", line 1, in <module>
    odo(data, pd.DataFrame)

  File "C:\Miniconda3\lib\site-packages\odo\odo.py", line 91, in odo
    return into(target, source, **kwargs)

  File "C:\Miniconda3\lib\site-packages\multipledispatch\dispatcher.py", line 164, in __call__
    return func(*args, **kwargs)

  File "C:\Miniconda3\lib\site-packages\blaze\compute\core.py", line 372, in into
    return into(a, b.data, **kwargs)

  File "C:\Miniconda3\lib\site-packages\multipledispatch\dispatcher.py", line 164, in __call__
    return func(*args, **kwargs)

  File "C:\Miniconda3\lib\site-packages\odo\into.py", line 43, in wrapped
    return f(*args, **kwargs)

  File "C:\Miniconda3\lib\site-packages\odo\into.py", line 53, in into_type
    return convert(a, b, dshape=dshape, **kwargs)

  File "C:\Miniconda3\lib\site-packages\odo\core.py", line 83, in __call__
    return _transform(self.graph, *args, **kwargs)

  File "C:\Miniconda3\lib\site-packages\odo\core.py", line 106, in _transform
    x = f(x, excluded_edges=excluded_edges, **kwargs)

  File "C:\Miniconda3\lib\site-packages\odo\backends\sql.py", line 747, in select_or_selectable_to_frame
    dtypes[field] = dtype.to_numpy_dtype()

  File "C:\Miniconda3\lib\site-packages\datashape\coretypes.py", line 911, in to_numpy_dtype
    return to_numpy_dtype(self)

  File "C:\Miniconda3\lib\site-packages\datashape\coretypes.py", line 1284, in to_numpy_dtype
    return to_numpy(ds.measure)[1]

  File "C:\Miniconda3\lib\site-packages\datashape\coretypes.py", line 1317, in to_numpy
    return tuple(shape), msr.to_numpy_dtype()

  File "C:\Miniconda3\lib\site-packages\datashape\coretypes.py", line 744, in to_numpy_dtype
    raise TypeError('DataShape measure %s is not NumPy-compatible' % self)

TypeError: DataShape measure ?int32 is not NumPy-compatible
```

**Comment by dhirschfeld** [comment]
With both commits odo can then correctly handle nullable foreign-keys:
```python
In [6]: data
Out[6]: <'Table' data; _name='_1', dshape='var * {  pkid: int32,  label_id: map[?int32, {pkid...'>

In [7]: odo(data, pd.DataFrame)
Out[7]: 
   pkid label_id
0     0        5
1     1        4
2     2        7
3     3        0
4     4        8
5     5        8
6     6        9
7     7        8
8     8        8
9     9     None

In [8]: _.dtypes
Out[8]: 
pkid         int32
label_id    object
dtype: object
```

**Comment by dhirschfeld** [comment]
@llllllllll - would be good to get this merged if you can find the time...

File: odo/backends/sql.py (hunk 2 above)
Comment (llllllllll) [inline odo/backends/sql.py]: I think we want to be strict about the dtypes to ensure that pandas isn't promoting things like crazy.

Comment (dhirschfeld) [inline odo/backends/sql.py]: I suspected as much which was why I left the numpy conversion inplace (sans the hard fail on unsupported types)

**Pull request merged**